keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -0,0 +1,223 @@
+- Stop defaulting component configs.
+- Outsource validation to the component binaries themselves.
+- Distinguish between user supplied and kubeadm generated and unmodified component configs.
+- Eradicate the need to convert configs
+- Use more community wide solutions, like Kustomize, kubeadm phases and addon operators
+
+### Non-Goals
+
+- Change the handling of kubeadm’s own config types in any way.
+- Cover in any way the migration of command line flags to component configs.
+- Define strategies for config map naming and backup.
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
+
+### Stop Component Config defaulting
+
+Component config defaulting has a number of issues:
+- It’s difficult to tell in retrospect if a setting value was intended or simply defaulted.
+- If a default value changes a defaulted setting will stop us from using the new value.
+- Defaulting bloats the resulting config, thus making it hard for users to migrate it upon a version change.
+
+Hence, kubeadm should stop defaulting component configs and leave this to components themselves.
+
+### Delegate config validation to the components
+
+Currently, kubeadm vendors in pieces of Go code from the components to be able to perform component config validation. This is a bad practice as it bloats the kubeadm dependencies and deviation between the vendored code and the components can lead to inconsistencies.  
+To solve this, validation is going to be performed by executing a container or the binary of the component to be used with an appropriate command to verify the supplied config.  
+If a target component does not have any means to verify its config, then kubeadm won’t do any validation for the time being. It will instead display a warning to the user denoting that certain component’s config is unvalidated.
+
+### Kubernetes Core vs AddOns Component Configs
+
+Currently kubeadm maintains the component configs of both kubelet and kube-proxy. These are components that are maintained in very different ways.
+
+Kubelet, along with API Server, Scheduler, Controller Manager and etcd, are deemed essential Kubernetes components. They are deployed and managed locally by kubeadm, using systemd services (kubelet) or static pods on the local kubelet (API Server, Scheduler, etc.). Hence, their configs are usually consumed by the components in the form of local files.
+
+On the other hand, kube-proxy and CoreDNS/kube-dns are addons, that are managed as standard Kubernetes deployments. Hence their configs are usually accessed via config maps.
+
+Therefore, two different config lifecycles are distinguishable here - one for Kubernetes core components and one for cluster addons.
+
+The defaulting and validation changes, proposed in the previous paragraphs, are going to be applied to both Kubernetes core and AddOn component config types.
+
+#### AddOn Component Config management
```

Comment:
does the `AddOn Component Config management` section need generated vs user supplied clarifications like the `Kubernetes Core Component Config management` section?

Reply:
It can surely make use of them. However, I intentionally kept it simple, so we can sync with cluster addon folks and come up with a viable solution.